refactor cached tensor more generic

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129359

# Motivation
solve https://github.com/pytorch/pytorch/issues/129027 to refactor cached tensor to be generic.

# Additional Context
No API name change. It is only decoupling with CUDA build option.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

guangyey committed:
refactor cached tensor more generic

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey commented:
@albanD could you help review this PR?

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -593,6 +594,57 @@ static PyObject* is_autocast_available(
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* set_cached_tensors_enabled(PyObject* _unused, PyObject* arg) {
```
Why did you choose to bind these by hand and not use pybind as in the previous code?
I would also suggest putting these in the main namespace and not autograd: https://github.com/pytorch/pytorch/blob/ca5d13c67287efe78a09dcf9b06f7e4fba898689/torch/csrc/Module.cpp#L1760

guangyey replied:
I think `cached tensor` is mainly used in autocast and autocast is under autogard namespace. So I put `cached tensor` to autograd namespace and bind them by hand. It is OK to me to move them to main namespace.

guangyey committed:
Update on "refactor cached tensor more generic"


# Motivation
solve https://github.com/pytorch/pytorch/issues/129027 to refactor cached tensor to be generic.

# Additional Context
No API name change. It is only decoupling with CUDA build option.

[ghstack-poisoned]

guangyey commented:
@pytorchbot rebase -b main

guangyey committed:
Update

[ghstack-poisoned]

guangyey commented:
@albanD May I know if I have addressed your comments?

albanD reviewed (approved):
That sounds ok. But as it stands, these API are private and so there is no BC guarantee for them. If you do plan on relying on them, I would suggest wrapping them somewhere in the torch.amp namespace.

guangyey commented:
> That sounds ok. But as it stands, these API are private and so there is no BC guarantee for them. If you do plan on relying on them, I would suggest wrapping them somewhere in the torch.amp namespace.
At this stage, It is enough. If these APIs are stable, I will export them to be public in a separate PR.

guangyey commented:
@pytorchbot merge
